WAF: Add a dedicated automatic-rules.php file for storing firewall rules

nateweller opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

This PR ensures that sites do not remove their automatic rules previously downloaded from WPCOM, in situations where the rules file is being re-generated and the site no longer has access to the latest firewall rules (i.e. cancelled their Scan plan).

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Adds a new `rules/automatic-rules.php` file for storing automatic rules fetched from WPCOM.
* Updates the generation of `rules/rules.php` to `require()` the relevant rules files, based on the WAF configuration (automatic and manual rules enabled options).
* Updates the Protect plugin to allow users to toggle automatic rules when they have the free plan but an old copy of `automatic-rules.php` available to use.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

peb6dq-p2-m9

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Start with a Jetpack site with a Scan plan.
* Activate the WAF module and ensure the `automatic-rules.php` file is generated, but not yet included in the main `rules.php` file.
* Validate a `?test=<script>` request is not blocked.
* Enable automatic rules.
* Validate a `?test=<script>` request is blocked.
* Enable manual rules and update the IP allow/block list with an IP address you can test.
* Validate the relevant IP rules file is updated, along with the main `rules.php` file.
* Validate the IP is blocked/allowed.
* Disable automatic rules.
* Validate that the relevant IP rules file is not updated, but no longer required by the main `rules.php` file.
* Validate the IP is no longer blocked/allowed.
* Cancel your Jetpack Scan plan.
* Toggle the WAF module off and then on again.
* Validate that the `automatic-rules.php` file still contains the automatic rules fetched at the beginning of your testing.

#### Screenshots

When the site previously upgraded and fetched automatic firewall rules, and then downgraded back to the free plan:

<img width="1042" alt="Screenshot 2023-01-04 at 11 30 39 AM" src="https://user-images.githubusercontent.com/10933065/210625120-8fd33110-e078-46f9-9fde-cb64218d9e2d.png">

nateweller committed:
Add a dedicated automatic-rules.php file for storing firewall rules fetched from WPCOM

nateweller committed:
Fix tests

nateweller committed:
Update tests

dkmyta commented:
With a fresh installation, using Protect, I was able to toggle the Automatic and Manual rules without any issue and the `rules.php` file was updated accordingly with requires for each of the corresponding `allow-ip.php`, `block-ip.php`, and `automatic-rules.php` files.

When canceling the plan, I could verify that the require for `automatic-rules.php` remained in the `rules.php` but that the content of the file was removed.

On the same install, via the Jetpack settings, the manual rules toggle handled similarly, however, regardless of the presence of a Scan plan, I was unable to get the `automatic-rules.php` require to apply when toggling the module.

miguelxpn commented on projects/packages/waf/src/class-waf-runner.php:
```diff
@@ -16,19 +16,21 @@
+	const IP_LISTS_ENABLED_OPTION_NAME             = 'jetpack_waf_ip_list';
+	const IP_ALLOW_LIST_OPTION_NAME                = 'jetpack_waf_ip_allow_list';
+	const IP_BLOCK_LIST_OPTION_NAME                = 'jetpack_waf_ip_block_list';
+	const RULES_FILE                               = __DIR__ . '/../rules/rules.php';
```
Maybe we should change the name? We have too many permutations of RULES_FILE naming and it's starting to get confused, maybe something like ENTRYPOINT_FILE?

nateweller replied:
I've organized the constants a bit, and renamed them as follows via https://github.com/Automattic/jetpack/pull/28027/commits/70f04f4f7af1ad968d26e82106eddc07e0f320e0:
* `RULES_ENTRYPOINT_FILE`
* `AUTOMATIC_RULES_FILE`
* `IP_ALLOW_RULES_FILE`
* `IP_BLOCK_RULES_FILE`

This introduces a naming system where the sub-rule file are all named `FOO_BAR_RULES_FILE` while the main file is `RULES_ENTRYPOINT_FILE` for extra distinction.

nateweller committed:
Rename rule file constants

nateweller committed:
Merge add/protect-waf-phase-2

nateweller committed:
Add default option for automatic rules enabled for backwards compat

nateweller committed:
Ensure activate hooks fire on insertion of options in addition to updates

nateweller commented:
> On the same install, via the Jetpack settings, the manual rules toggle handled similarly, however, regardless of the presence of a Scan plan, I was unable to get the `automatic-rules.php` require to apply when toggling the module.

I pulled in some backwards compatibility changes from https://github.com/Automattic/jetpack/pull/27785 to resolve this issue (https://github.com/Automattic/jetpack/pull/28027/commits/d9ae22eb44903e5109adcbb386f085ef341dd858 and https://github.com/Automattic/jetpack/pull/28027/commits/0da59ac6d8bfd5d0c2856b77d8a0b98956ae1490).

This PR is now ready for re-review :+1:

nateweller committed:
Remove unrelated changes from merge

nateweller committed:
Merge branch 'add/protect-waf-phase-2' into add/waf-automatic-rules-file

nateweller committed:
Changelog

nateweller committed:
Changelog

nateweller committed:
Merge branch 'add/protect-waf-phase-2' into add/waf-automatic-rules-file

dkmyta commented:
When I first activated Protect on a fresh Jurassic Tube instance, the Firewall was disabled by default - after hitting the prompt to enable, the notice remained after the loading icon resolved, prompting me again to enable it until I refreshed the page. The WAF module did appear to be enabled on the initial click and the endpoints were registered, however, it seems that the latest changes may have introduced an issue with toggling the Automatic and Manual rules settings. The initial attempt presents an error notice, and an immediate second attempt succeeds. A refresh of the page on the initial attempt displays as if the settings are active, however, in either case (regardless of a plan), no rules file is generated, no rules are run, and we are unable to successfully save changes to the manual rules settings, although we can access them. 

With the WAF module enabled and all other settings off, activating Jetpack produces the following fatal error because we never initially generated the rules file. It is not clear to me why `Waf_Runner::deactivate` is run since I was only enabling Jetpack, not deactivating Protect or any other WAF settings:
```
PHP Fatal error:  Uncaught Exception: Failed to empty rules.php file. in /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-runner.php:403
Stack trace:
#0 /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-initializer.php(54): Automattic\Jetpack\Waf\Waf_Runner::deactivate()
#1 /var/www/html/wp-includes/class-wp-hook.php(308): Automattic\Jetpack\Waf\Waf_Initializer::on_deactivation()
#2 /var/www/html/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters()
#3 /var/www/html/wp-includes/plugin.php(517): WP_Hook->do_action()
#4 /usr/local/src/jetpack-monorepo/projects/packages/status/src/class-modules.php(586): do_action()
#5 /usr/local/src/jetpack-monorepo/projects/plugins/jetpack/class.jetpack.php(592): Automattic\Jetpack\Modules->update_active()
#6 /usr/local/src/jetpack-monorepo/projects/plugins/jetpack/class.jetpack.php(601): Jetpack::update_active_modules()
#7 /usr/local/src/jetpack-monorepo/projects/plugins/jetpack/class.jetpack.php(2919): Jetpack::delete_active_modules()
#8 /usr/local/src/jetpack-monorepo/projects/plugins/jetpack/class.jetpack.php(2898): Jetpack::handle_default_module_activation()
#9 /usr/local/src/jetpack-monorepo/projects/plugins/jetpack/class.jetpack.php(2783): Jetpack::plugin_initialize()
#10 /var/www/html/wp-includes/class-wp-hook.php(308): Jetpack::plugin_activation()
#11 /var/www/html/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters()
#12 /var/www/html/wp-includes/plugin.php(517): WP_Hook->do_action()
#13 /var/www/html/wp-admin/includes/plugin.php(691): do_action()
#14 /var/www/html/wp-admin/plugins.php(58): activate_plugin()
#15 {main}
  thrown in /usr/local/src/jetpack-monorepo/projects/packages/waf/src/class-waf-runner.php on line 403
  ```

nateweller committed:
Merge branch 'add/protect-waf-phase-2' into add/waf-automatic-rules-file

nateweller committed:
Assorted fixes missed during merge resolution

nateweller committed:
Fix module activation when on a site level connection

nateweller committed:
[not verified] Restore masterbar file that was accidentally formatted

nateweller committed:
Merge branch 'add/protect-waf-phase-2' into add/waf-automatic-rules-file

nateweller committed:
Refresh WAF data after connecting the site

nateweller committed:
Allow sites with old firewall rule files available to toggle automatic rules, update related messaging

nateweller commented:
@dkmyta Thanks so much for all of the testing and assistance on this PR! 

I've updated with all of the latest changes from trunk/phase 2, and fixed the issues you were having. Now, the Protect UI will allow you to manage automatic rules after downgrading.

Ready for your re-review 🙏

nateweller committed:
Fix translation minification errors

nateweller committed:
Additional translation fixes

nateweller commented:
Note: Changelogger validity will be fixed in the [add/protect-waf-phase-2](https://github.com/Automattic/jetpack/tree/add/protect-waf-phase-2) branch.

dkmyta reviewed (approved):
Works great!
